File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/ScreenSwitcherApplication.java
package com.jaynewstrom.screenswitchersample;

import android.app.Application;

import com.jaynewstrom.concrete.Concrete;
import com.jaynewstrom.concrete.ConcreteWall;
import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;

import timber.log.Timber;

public final class ScreenSwitcherApplication extends Application {

    private ConcreteWall<ApplicationComponent> foundation;
    private RefWatcher refWatcher;

    @Override public void onCreate() {
        super.onCreate();
        Timber.plant(new Timber.DebugTree());
        refWatcher = LeakCanary.install(this);
        foundation = Concrete.pourFoundation(applicationComponent());
    }

    private ApplicationComponent applicationComponent() {
        return DaggerApplicationComponent.builder().applicationModule(new ApplicationModule(this, refWatcher)).build();
    }

    @Override public Object getSystemService(String name) {
        if (Concrete.isService(name)) {
            return foundation;
        }
        return super.getSystemService(name);
    }
}

File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/ScreenSwitcherApplication.kt
package com.jaynewstrom.screenswitchersample

import android.app.Application
import android.content.Context

import com.jaynewstrom.concrete.Concrete
import com.jaynewstrom.concrete.ConcreteWall
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher

import timber.log.Timber

class ScreenSwitcherApplication : Application() {
    private lateinit var foundation: ConcreteWall<ApplicationComponent>
    private lateinit var refWatcher: RefWatcher

    override fun onCreate() {
        super.onCreate()
        Timber.plant(Timber.DebugTree())
        refWatcher = LeakCanary.install(this)
        foundation = Concrete.pourFoundation(applicationComponent())
    }

    private fun applicationComponent(): ApplicationComponent {
        return DaggerApplicationComponent.builder().build()
    }

    override fun getSystemService(name: String): Any? {
        return if (Concrete.isService(name)) {
            foundation
        } else super.getSystemService(name)
    }

    companion object {
        fun watchObject(context: Context, watchedReference: Any) {
            val application = context.applicationContext as ScreenSwitcherApplication
            application.refWatcher.watch(watchedReference)
        }
    }
}
